src/java.base/share/classes/jdk/internal/module/ModuleBootstrap.java

```diff
@@ -469,15 +468,27 @@ private static ModuleLayer boot2() {
                 limitedFinder = new SafeModuleFinder(finder);
         }
 
+        // If -Xshare:dump and mainModule are specified, check if the mainModule
+        // is in the runtime image and not in the upgrade module path. If so,
+        // set canArchive to true so that the module graph can be archived.
+        if (CDS.isDumpingStaticArchive() && mainModule != null) {
+            ModuleReference mainModuleRef = systemModuleFinder.find(mainModule).orElse(null);
+            if (mainModuleRef != null) {
+                URI mainModuleLocation = mainModuleRef.location().orElse(null);
+                if (mainModuleLocation != null && mainModuleLocation.getScheme().equalsIgnoreCase("jrt")) {
+                    canArchive = true;
+                }
+            }
```

Comment:
I tried the above but got the following build error:
```
Optimizing the exploded image
Error occurred during initialization of boot layer
java.lang.NullPointerException
ExplodedImageOptimize.gmk:39: recipe for target '/scratch/cccheung/ws/git/curr_jdk/build/mywork-debug/jdk/_optimize_image_exec.marker' failed
```
Let's keep the existing code for now because I may need to refactor the code into a utility method for a upcoming RFE to support the --add-modules option.

Reply:
> I tried the above but got the following build error:
> 
> ```
> Optimizing the exploded image
> Error occurred during initialization of boot layer
> java.lang.NullPointerException
> ExplodedImageOptimize.gmk:39: recipe for target '/scratch/cccheung/ws/git/curr_jdk/build/mywork-debug/jdk/_optimize_image_exec.marker' failed
> ```

I checked tier1 with this change and don't see an issue, is it possible that something was left over in your env?